main/acle.md

```diff
@@ -4932,6 +5044,1381 @@ assertions, which may speed up access to aligned data (and will fault
+functions do. See [[AAPCS64]](#AAPCS64) for more details about
+private-ZA interfaces.
+
+A function definition with this attribute is [ill-formed](#ill-formed)
+if the function's type has an [`arm_shared_za`](#arm_shared_za)
+attribute or an [`arm_preserves_za`](#arm_preserves_za) attribute.
+
+#### `arm_preserves_za`
+
+This attribute applies to **function types** and is simply an optimization
+hint to the compiler; it is never needed for correctness. It can be attached
+to any function type, including:
+
+*   a function type that has an [`arm_shared_za`](#arm_shared_za) attribute
+
+*   the type of a function whose definition has an [`arm_new_za`](#arm_new_za)
+    attribute
+
+*   the type of a function whose definition has no ZA state
+
+The attribute specifies that the function “preserves ZA”, in the sense
+of [[AAPCS64]](#AAPCS64). The mapping of this PCS concept to C and C++
+depends on whether the function is [shared-ZA](#shared-za) or
+[private-ZA](#private-za):
+
+*   If the function is [shared-ZA](#shared-za), the attribute guarantees
+    that the contents of the shared [ZA state](#za-state) on return from
+    the function are the same as the contents of the shared ZA state on
+    entry to the function. Either the function does not change the ZA
+    state at all, or the function undoes any changes to the ZA state
+    before returning.
+
+    Note that C and C++ calls from shared-ZA functions to private-ZA
+    functions are defined to preserve ZA state. Such calls do not affect
+    whether a function can be marked [`arm_preserves_za`](#arm_preserves_za).
+
+*   If the function is [private-ZA](#private-za), the attribute guarantees
+    that a call to the function does not directly or indirectly involve
+    any of the following:
+
+    *   a call to an [`arm_new_za`](#arm_new_za) function
+
+    *   a call to `setjmp`
+
+    *   a call to [`__arm_za_disable`](#__arm_za_disable)
+
+    *   a call to a non-C/C++ function that “commits a lazy save”, in the
+        sense of the [[APCS64]](#AAPCS64)
+
+    *   an inline asm that commits a lazy save
+
+    *   the catching of an exception
+
+    The platform may place additional requirements as well.
+
+In both cases, the onus is on the definition of the function to honor
+the guarantee that is being made. The attribute does not direct the
+compiler to do anything to honor the guarantee.
+
+If a function with an `arm_preserves_za` type does not preserve ZA,
+the behavior is undefined. (There is an analogy with functions
+that are declared `noreturn` but do in fact return, and to functions
+that are declared `const` but do in fact change memory.)
+
+The attribute is intended to be useful for functions at API boundaries,
+where the compiler might not be able to “see” the definition of the
+function being called. As the description above implies, attaching
+`arm_preserved_za` to a [private-ZA](#private-za) function is quite
```

Comment:
What's the difference between a low level feature and a quite low level feature? :)
```suggestion
`arm_preserved_za` to a [private-ZA](#private-za) function is
```

Reply:
How about “rather a low-level feature”?  I'd prefer to leave some equivocation in there since to me “low-level feature” implies something that should not be used in general code.  I don't think that applies here.

Reply:
Oh, I see, `quite a low-level feature` works well because it is follower by `..., but it is useful`. I think we can leave this one as it is. I'll mark the conversation as resolved.